apps/src/javalab/Javalab.js

```diff
@@ -120,7 +139,117 @@ Javalab.prototype.onContinue = function() {
 };
 
 Javalab.prototype.getCode = function() {
-  return '';
+  // store the file version as the source, as we do in WebLab
+  return this.getCurrentFilesVersionId();
+};
+
+Javalab.prototype.getCurrentFilesVersionId = function() {
+  return project.filesVersionId || this.initialFilesVersionId;
+};
+
+// Called by Javalab when a file has been renamed
+Javalab.prototype.renameProjectFile = function(filename, newFilename) {
+  // no need to rename if the name hasn't changed
+  if (filename === newFilename) {
+    return;
+  }
+  header.showProjectSaving();
+  filesApi.renameFile(
+    filename,
+    newFilename,
+    () => {
+      // at this point project.filesVersionId will be updated and we can save
+      // sources as well.
+      project.save();
+    },
+    () => {
+      console.warn(`Javalab: error file ${filename} not renamed`);
+      project.showSaveError_();
+    }
+  );
 };
 
+Javalab.prototype.onSave = function() {
+  header.showProjectSaving();
+  const filename = getStore().getState().javalab.filename;
+  const editorText = getStore().getState().javalab.editorText;
+  // TODO: enable multi-file
+  filesApi.putFile(
```

Comment:
Should this call `saveIfSourcesChanged`?

Reply:
it calls `save`, which doesn't check if the sources changed. The sources here is the files version id, which I think will always change. If we prefer `saveIfSourcesChanged` I can change it to that.

Reply:
That makes sense. I forgot that this is the files api, not the sources api